### PR TITLE
[Final] Handle UTF-8 encoding via major updates to how lua strings work

### PR DESCRIPTION
**Pull request by jpoka**
This PR changes how lua strings work in a major way. Now, indexing into strings returns the ith character instead of the ith byte, `string.len(s)` and `#s` return character counts instead of byte counts, and so on. Open source code came from https://github.com/starwing/luautf8.

### CODE REVIEW

File: 3rdparty/lua-5.1.3/src/lvm.c [hunk 1]
```diff
@@ -515,7 +515,24 @@ void luaV_execute (lua_State *L, int nexeccalls) {
...
           }
           case LUA_TSTRING: {
-            setnvalue(ra, cast_num(tsvalue(rb)->len));
+            const char *bytes = getstr(tsvalue(rb));
+            size_t index = 0;
+            size_t characterLength = 0;
+            unsigned char byte = bytes[index];
+            while (byte) {
+              size_t byteCount = 1;
+              if (byte >= 240) {
```
Comment (stevenmah) [inline 3rdparty/lua-5.1.3/src/lvm.c]: Disclaimer: I'm not super familiar with UTF-8 encodings, so all this info is pulled from the [wiki entry](https://en.wikipedia.org/wiki/UTF-8)

With that in mind and according to the table on that page, there are a few restricted starting bytes that are always invalid and some starting bytes that are valid _some_ of the time, but not all the time. Not sure if we should be more specific in this if-else block or not to accommodate that.

That said, ahu tests in `wonder_content` seem to be passing, so I'm not sure what to make of it. In any case, perhaps we can move these comparison numbers into constants in the file.

Comment (jpoka) [reply 3rdparty/lua-5.1.3/src/lvm.c]: It's a good question. I don't know the answer yet. Also, this code assumes a valid UTF-8 encoding. That's probably a safe assumption for the data we control but not for anything provided by the user. Note the above code is for the string length `#` operator.

Comment (jpoka) [reply 3rdparty/lua-5.1.3/src/lvm.c]: Updated string length operator to be more strict.

**Comment by stevenmah** [comment]
:shipit: Update looks good to me. Seems the "sometimes valid, sometimes invalid" category of leading bytes apply to Korean and Indic characters, and a few other multilingual categories (like characters used in Chinese, Japanese, and Korean). I think it's alright to ignore those given that, for now at least, we aren't tackling those areas for localization.

**Comment by antonmorant** [comment]
Looks good to me!
:shipit:

**Pull request merged**